hledger: update livecheck

samford opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

#66319 was merged before I had a chance to review it, so this PR contains comments/changes I would have proposed in review.

The previous `livecheck` block for `hledger` only contained `url :stable`, so livecheck was using the `Hackage` strategy to check for new versions. livecheck was erroneously reporting `1.20` as the newest version but `1.19.1` is the current version on the [first-party download page](https://hledger.org/download.html), [GitHub](https://github.com/simonmichael/hledger/releases/latest), etc. Since a new version may be added to Hackage before it's officially released, it's necessary for us to check a different source to find the current version.

Though the "latest" release on GitHub corresponds to the current release on the first-party download page at the present moment, it may be slightly more appropriate to check the first-party download page itself. My thinking here is that the first-party download page may be a more canonical representation of when a version is officially released. I've seen the "latest" version on GitHub for some other projects point to incorrect releases at times, so I think the first-party download page has the potential to be slightly less error-prone in the long run.

---

That said, regardless of whether we modify the `url` in the `livecheck` block, I think it's important to have a comment to explain why it's necessary to check a source that doesn't align with the `stable` URL for this formula. If we don't end up wanting to modify the `url`, I'll update the language in the comment accordingly.

chenrui333 commented:
> That said, regardless of whether we modify the url in the livecheck block, I think it's important to have a comment to explain why it's necessary to check a source that doesn't align with the stable URL for this formula. If we don't end up wanting to modify the url, I'll update the language in the comment accordingly.

+1

chenrui333 commented:
yeah, I did not bother checking the download page. 😅 

I wonder if we can update docs (if exits) about the preference order between `download page`, `ftp`(common registry), `github tags`.

samford commented:
I'm working on addressing feedback in my Homebrew/brew documentation PR, so I'll see if I can add this information somewhere (e.g., the URL guidelines).

In other formulae, we generally check the first-party website when the developers have a habit of uploading artifacts for a version before it's officially released (the check in this PR follows the same pattern). Though this is common, it depends on how versions are announced for a particular formula, so I think we have to be careful about what we suggest.

simonmichael commented:
Thanks for working on this. Indeed sometimes problems cause time lag between the various locations (as with the 1.20 release). I tend to agree that the download page is the best place to check, specifically the sentence that currently reads "The current hledger release is 1.20" at the top.
